main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java

```diff
@@ -47,32 +58,100 @@
+      // this trip_id
+      List<StopTooFarFromTripShapeNotice> noticesForTrip =
+          checkStopsWithinTripShape(
+              tripId,
+              stopTimesForTrip,
+              trip.shapeId(),
+              shapeTable.byShapeId(trip.shapeId()),
+              stopTable,
+              testedCache);
+      for (StopTooFarFromTripShapeNotice notice : noticesForTrip) {
+        noticeContainer.addValidationNotice(notice);
+      }
+    }
+  }
+
+  /**
+   * Returns a list of notices for the given input, one for each stop that is too far from the trip
+   * shapePoints
+   *
+   * @param tripId trip_id for this GTFS trip
+   * @param stopTimes a list of StopTimes for a trip, sorted by stop_sequence
+   * @param shapeId the shape_id for this GTFS trip
+   * @param shapePoints a list of ShapePoints for a trip, sorted by shape_pt_sequence
+   * @param stopTable a container for all stops (keyed on stop_id), needed to obtain the latitude
+   *     and longitude for each stop on the trip
+   * @param testedCache a cache for previously tested shape_id and stop_id pairs (keyed on
+   *     shape_id+stop_id). If the combination of shape_id and stop_id appears in this set, we
+   *     shouldn't test it again. Shapes and stops tested in this method execution will be added to
+   *     this testedCache.
+   * @return a list of notices, one for each stop that is too far from the trip shape
+   */
+  List<StopTooFarFromTripShapeNotice> checkStopsWithinTripShape(
+      final String tripId,
+      final List<GtfsStopTime> stopTimes,
+      final String shapeId,
+      final List<GtfsShape> shapePoints,
+      final GtfsStopTableContainer stopTable,
+      final Set<String> testedCache) {
+    List<StopTooFarFromTripShapeNotice> notices = new ArrayList<>();
+    if (stopTimes == null || stopTimes.isEmpty() || shapePoints == null || shapePoints.isEmpty()) {
+      // Nothing to do - return empty list
+      return notices;
+    }
+
+    // Create a polyline from the GTFS shapes data
+    ShapeFactory.LineStringBuilder lineBuilder = GeospatialUtil.getShapeFactory().lineString();
+    for (GtfsShape shapePoint : shapePoints) {
+      lineBuilder.pointXY(shapePoint.shapePtLon(), shapePoint.shapePtLat());
+    }
+    Shape shapeLine = lineBuilder.build();
+
+    // Create the buffered version of the trip as a polygon
+    Shape shapeBuffer = shapeLine.getBuffered(TRIP_BUFFER_DEGREES, shapeLine.getContext());
+
+    // Check if each stop is within the buffer polygon
+    for (GtfsStopTime stopTime : stopTimes) {
+      GtfsStop stop = stopTable.byStopId(stopTime.stopId());
+      if (stop == null || !stop.hasStopLat() || !stop.hasStopLon()) {
+        // Lat/lon are optional for location_type 4 - skip to the next stop if they aren't
+        // provided
+        continue;
+      }
+      if (!(stop.locationType() == GtfsLocationType.STOP)
+          && !(stop.locationType() == GtfsLocationType.BOARDING_AREA)) {
```

Comment:
This can be written simpler:

if (stop.locationType() != GtfsLocationType.STOP &&
stop.locationType() != GtfsLocationType.BOARDING_AREA)

However, there is already a requirement that a trip must visit only stops or boarding areas, so I think you can omit that extra check.

Reply:
>However, there is already a requirement that a trip must visit only stops or boarding areas, so I think you can omit that extra check.

@aababilov Is this checked in the existing validator? I'm not seeing it, but could be missing it.

Reply:
I went ahead and removed the check here in https://github.com/MobilityData/gtfs-validator/pull/750/commits/b6c414917461c0d1951ac7223691555b71e7e8c7 as I agree that it should be checked elsewhere.

Reply:
I've opened this issue to add checks for location_types of stop_times.stop_id stops here - https://github.com/MobilityData/gtfs-validator/issues/763.